### PR TITLE
8334421: assert(!oldbox->is_unbalanced()) failed: this should not be called for unbalanced region

### PR DESCRIPTION
**Pull request by vnkozlov**
[JDK-8324969](https://bugs.openjdk.org/browse/JDK-8324969) changes was not complete. When Escape Analysis tries to determine if lock can be eliminated it looks only if corresponding object does not escape. I does not take into account if coarsening locks elimination already happened and created unbalanced regions.

Added check that not-escaped object is not referenced from an unbalanced region before marking or eliminating corresponding locks.

Dump additional information for coarsened locks when debugging.

New regression test was added

Tested tier1-5,stress,xcomp

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334421](https://bugs.openjdk.org/browse/JDK-8334421): assert(!oldbox-&gt;is_unbalanced()) failed: this should not be called for unbalanced region (**Bug** - P3)

### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [312f59f4](https://git.openjdk.org/jdk/pull/19870/files/312f59f4955a797fa26f50155da4cea003c40802)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [312f59f4](https://git.openjdk.org/jdk/pull/19870/files/312f59f4955a797fa26f50155da4cea003c40802)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19870/head:pull/19870` \
`$ git checkout pull/19870`

Update a local copy of the PR: \
`$ git checkout pull/19870` \
`$ git pull https://git.openjdk.org/jdk.git pull/19870/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19870`

View PR using the GUI difftool: \
`$ git pr show -t 19870`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19870.diff">https://git.openjdk.org/jdk/pull/19870.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19870#issuecomment-2187598935)

### CODE REVIEW

File: src/hotspot/share/opto/escape.cpp [hunk 1]
```diff
@@ -3503,7 +3503,7 @@ bool ConnectionGraph::not_global_escape(Node *n) {
 // all compiled code paths have corresponding Lock/Unlock pairs.
 bool ConnectionGraph::can_eliminate_lock(AbstractLockNode* alock) {
   BoxLockNode* box = alock->box_node()->as_BoxLock();
```
Comment (iwanowww) [inline src/hotspot/share/opto/escape.cpp:3505]: The definition can be moved under `EliminateNestedLocks` guard.

**Review by iwanowww — APPROVED** [review]
Looks good.

File: src/hotspot/share/opto/escape.cpp (hunk 1 above)
Comment (vnkozlov) [reply src/hotspot/share/opto/escape.cpp:3505]: Done.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me too.

**Comment by vnkozlov** [comment]
Thank you, Vladimir and Tobias, for reviews.

**Pull request closed without merging**